src/macaron/provenance/provenance_finder.py

```diff
@@ -490,3 +499,96 @@ def download_provenances_from_ci_service(ci_info: CIInfo, download_path: str) ->
+) -> str | None:
+    """Get the hash of the artifact found from the passed PURL using local or remote files.
+
+    Provided local caches will be searched first. Artifacts will be downloaded if nothing is found within local
+    caches, or if no appropriate cache is provided for the target language.
+    Downloaded artifacts will be added to the passed package registry to prevent downloading them again.
+
+    Parameters
+    ----------
+    purl: PackageURL
+        The PURL of the artifact.
+    local_artifact_dirs: list[str] | None
+        The list of directories that may contain the artifact file.
+    package_registries_info: list[PackageRegistryInfo]
+        The list of package registry information.
+
+    Returns
+    -------
+    str | None
+        The hash of the artifact, or None if no artifact can be found locally or remotely.
+    """
+    if local_artifact_dirs:
+        # Try to get the hash from a local file.
+        artifact_hash = get_local_artifact_hash(purl, local_artifact_dirs)
+
+        if artifact_hash:
+            return artifact_hash
+
+    # Download the artifact.
+    if purl.type == "maven":
+        maven_registry = next(
+            (
+                package_registry
+                for package_registry in PACKAGE_REGISTRIES
+                if isinstance(package_registry, MavenCentralRegistry)
+            ),
+            None,
+        )
+        if not maven_registry:
+            return None
+
+        return maven_registry.get_artifact_hash(purl)
+
+    if purl.type == "pypi":
+        pypi_registry = next(
+            (package_registry for package_registry in PACKAGE_REGISTRIES if isinstance(package_registry, PyPIRegistry)),
+            None,
+        )
+        if not pypi_registry:
+            logger.debug("Missing registry for PyPI")
+            return None
+
+        registry_info = next(
+            (
+                info
+                for info in package_registries_info
+                if info.package_registry == pypi_registry and info.build_tool_name in {"pip", "poetry"}
+            ),
+            None,
+        )
+        if not registry_info:
+            logger.debug("Missing registry information for PyPI")
+            return None
+
+        if not purl.version:
+            return None
+
+        pypi_asset = find_or_create_pypi_asset(purl.name, purl.version, registry_info)
+        if not pypi_asset:
+            return None
+
+        pypi_asset.has_repository = True
+        if not pypi_asset.download(""):
+            return None
+
+        artifact_hash = pypi_asset.get_sha256()
+        if artifact_hash:
+            return artifact_hash
+
+        source_url = pypi_asset.get_sourcecode_url("bdist_wheel")
+        if not source_url:
+            return None
+
+        return pypi_registry.get_artifact_hash(source_url)
+
+    logger.debug("Purl type '%s' not yet supported for GitHub attestation discovery.", purl.type)
```

Comment:
I wonder if we should change this debug message as this function doesn't handle attestation?

Reply:
Technically this function is only used for GitHub attestation purposes at the moment, but I agree it should be changed.